apps/aecore/src/aec_chain_state.erl

```diff
@@ -0,0 +1,458 @@
+
+wrap_block(Block) ->
+    {ok, Hash} = aec_blocks:hash_internal_representation(Block),
+    #node{ type = block
+         , content = Block
+         , difficulty = aec_blocks:difficulty(Block)
+         , hash = Hash
+         }.
+
+wrap_header(Header) ->
+    {ok, Hash} = aec_headers:hash_header(Header),
+    #node{ type = header
+         , content = Header
+         , difficulty = aec_headers:difficulty(Header)
+         , hash = Hash
+         }.
+
+export_header(#node{type = header, content = H}) -> H;
+export_header(#node{type = block, content = H}) -> aec_blocks:to_header(H).
+
+export_block(#node{type = block, hash = H, content = B}, State) ->
+    case add_state_tree_to_block(B, H, State) of
+        {ok, ExportBlock} -> ExportBlock;
+        error -> B
+    end.
+
+%%%-------------------------------------------------------------------
+%%% Handling difficulty
+%%%-------------------------------------------------------------------
+
+total_difficulty_at_hash(Hash, State) ->
+    total_difficulty_at_hash(Hash, 0, State).
+
+total_difficulty_at_hash(Hash, Acc, State) ->
+    case blocks_db_find(Hash, State) of
+        {ok, Node} ->
+            NewAcc = Acc + node_difficulty(Node),
+            PrevHash = prev_hash(Node),
+            case node_is_genesis(Node) of
+                true -> {ok, NewAcc};
+                false -> total_difficulty_at_hash(PrevHash, NewAcc, State)
+            end;
+        error ->
+            {error, not_rooted}
+    end.
+
+%%%-------------------------------------------------------------------
+%%% Handling the state trees
+%%%-------------------------------------------------------------------
+
+add_state_tree_to_block(Block, Hash, State) ->
+    case state_db_find(Hash, State) of
+        {ok, Trees} -> {ok, aec_blocks:set_trees(Block, Trees)};
+        error -> error
+    end.
+
+%%%-------------------------------------------------------------------
+%%% Chain operations
+%%%-------------------------------------------------------------------
+
+internal_insert(Node, State) ->
+    case blocks_db_find(Node#node.hash, State) of
+        error -> check_update_after_insert(Node, blocks_db_put(Node, State));
+        {ok, Node} -> State;
+        {ok, Old} when Node#node.type =:= block, Old#node.type =:= header ->
+            check_update_after_insert(Node, blocks_db_put(Node, State));
+        {ok, Old} when Node#node.type =:= Old#node.type ->
+            internal_error({same_key_different_content, Node, Old});
+        {ok, Old} when Node#node.type =:= header, Old#node.type =:= block ->
+            case export_header(Old) =:= export_header(Node) of
+                true  -> State;
+                false -> internal_error({different_header_existing_block, Old, Node})
+            end
+    end.
+
+check_update_after_insert(Node, State) ->
+    case determine_chain_relation(Node, State) of
+        off_chain -> State;
+        in_chain -> update_state_tree(Node, State);
+        {fork, ForkNode} ->
```

Comment:
please confirm: here we switch over to fork always when we get it.

We need to add difficulty check for incoming fork (just to make sure if someone is not cheating with lower target). Actually, we need to compute total difficulty in each branch from fork point

Reply:
Yes, this is done. Look at determine_new_header_top.

Reply:
I will make this more clear in the code.